crates/ark/src/analysis/parse_boundaries.rs

```diff
@@ -170,190 +200,200 @@ fn merge_overlapping(ranges: Vec<LineRange>) -> Vec<LineRange> {
 // Fill any gaps with one-liner complete expressions. This applies
 // to empty lines, with or without spaces and tabs, and possibly
 // with a trailing comment.
-fn fill_gaps(mut boundaries: ParseBoundaries, n_lines: u32) -> ParseBoundaries {
-    let mut filled = vec![];
-    let ranges = boundaries.complete;
-
+fn fill_gaps(
+    complete: Vec<LineRange>,
+    incomplete: &Option<LineRange>,
+    invalid: &Option<LineRange>,
+    n_lines: u32,
+) -> Vec<ParseBoundary> {
+    let mut filled: Vec<ParseBoundary> = vec![];
     let mut last_line: u32 = 0;
 
     let range_from = |start| LineRange::new(start, start + 1);
 
     // Fill leading whitespace with empty input ranges
-    if let Some(first) = ranges
-        .get(0)
-        .or(boundaries.incomplete.as_ref())
-        .or(boundaries.error.as_ref())
-    {
+    if let Some(first) = complete.get(0).or(incomplete.as_ref()).or(invalid.as_ref()) {
         for start in 0..first.start() {
             let range = range_from(start);
             last_line = range.end();
-            filled.push(range)
+            filled.push(ParseBoundary::whitespace(range))
         }
     }
 
     // Fill gaps between complete expressions
-    for range in ranges.into_iter() {
+    for range in complete.into_iter() {
         // We found a gap, fill ranges for lines in that gap
         if !range.contains(last_line) {
             for start in last_line..range.start() {
-                filled.push(range_from(start))
+                filled.push(ParseBoundary::whitespace(range_from(start)))
             }
         }
 
         last_line = range.end();
-        filled.push(range);
+        filled.push(ParseBoundary::complete(range));
     }
 
     // Fill trailing whitespace between complete expressions and the rest
-    // (incomplete, error, or eof)
-    let last_boundary = filled.last().map(|r| r.end()).unwrap_or(0);
-    let next_boundary = boundaries
-        .incomplete
+    // (incomplete, invalid, or eof)
+    let last_boundary = filled.last().map(|b| b.range.end()).unwrap_or(0);
+    let next_boundary = incomplete
         .as_ref()
-        .or(boundaries.error.as_ref())
+        .or(invalid.as_ref())
         .map(|r| r.start())
         .unwrap_or(n_lines);
 
     for start in last_boundary..next_boundary {
```

Comment:
```suggestion
    for start in last_complete_boundary..next_boundary {
```

Maybe? Having `last_boundary..next_boundary` is mildly confusing because "last" implies its the...last one, but then we are saying there is a "next" one after that